[SPARK-8743] [Streaming]: Deregister Codahale metrics for streaming when StreamingContext is closed

nssalian opened the pull request:
The issue link: https://issues.apache.org/jira/browse/SPARK-8743
Deregister Codahale metrics for streaming when StreamingContext is closed

Design:
Adding the method calls in the appropriate start() and stop () methods for the StreamingContext

Actions in the PullRequest:
1) Added the registerSource method call to the start method for the Streaming Context. 
2) Added the removeSource method to the stop method. 
3) Added comments for both 1 and 2 and comment to show initialization of the StreamingSource

Requesting Review.

Commit:
SPARK-8743: Added the registerSource method call to the start method for the Streaming Context. Added the removeSource method to the stop method. Added comments for both

Added // instead of /** for commenting in code

Added indentation and Space at the comment on line 578; Registering..

Added test case for de-register metrics and made a change to the scope of the sources ArrayBuffer

Added additional variable to check the updated Sources size value to compare with the original size after removal

Added the removeSource method in try

Removed the assert for the env field, added the registerSource line in the INITIALIZED block and kept the removeSource() in the ACTIVE block

Added test to check registering and de-registering of streamingSource

Removed unused imports

Moved the registerSource() call before line 601

Changed scope of sources and corrected comments for helper

Removed extra line in Helper Methods section

Added helper method for private methods and changed the test logic to check for Sources containing or not containing StreamingSource

Changed import statements to remove unnecessary imports and add specific imports

Removed types for fields in test for registering and deregistering metrics

Changed imports statements, negated test statement and removed postfix

Removed added comment to Assert for INITIALIZED state

Removing the INITIALIZED check since after start() the state moves to ACTIVE and this check fails

Move the INITIALIZED state check to when the ssc is initialized

Commit:
SPARK-8743: Added the registerSource method call to the start method for the Streaming Context. Added the removeSource method to the stop method. Added comments for both

AmplabJenkins commented:
Can one of the admins verify this patch?

jerryshao commented on streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala:
```diff
@@ -577,6 +575,12 @@ class StreamingContext private[streaming] (
    */
   def start(): Unit = synchronized {
+  /**
+   * Registering Streaming Metrics at the start of the StreamingContext
```
I think it is better to use `//` rather than `/** */` for internal code commenting.

jerryshao commented:
This change LGTM, actually will get rid of some annoying logs :).

Commit:
Added // instead of /** for commenting in code

nssalian commented:
@jerryshao thank you for the comment. I made the changes.
Please let me know if you think I could add anything additional.

srowen commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36660 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36660/consoleFull) for   PR 7250 at commit [`a665965`](https://github.com/apache/spark/commit/a665965a40c7b49cc13d8ab38f8da194693d9845).

SparkQA commented:
[Test build #36660 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36660/console) for   PR 7250 at commit [`a665965`](https://github.com/apache/spark/commit/a665965a40c7b49cc13d8ab38f8da194693d9845).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

srowen commented on streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala:
```diff
@@ -577,6 +575,10 @@ class StreamingContext private[streaming] (
    * @throws IllegalStateException if the StreamingContext is already stopped.
    */
   def start(): Unit = synchronized {
+  //Registering Streaming Metrics at the start of the StreamingContext
```
My guess is this fails style checks. Indent it, and add a space

Commit:
Added indentation and Space at the comment on line 578; Registering..

nssalian commented:
@srowen, I made the changes and ran a dev test on my repo. The scala errors weren't present.

srowen commented:
@nssalian that is indeed the failure: "Insert a space after the start of the comment" Did you run `./dev/scalastyle`?

nssalian commented:
nsalian-MBP:spark nsalian$ ./dev/scalastyle
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=512m; support was removed in 8.0
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=512m; support was removed in 8.0
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=512m; support was removed in 8.0
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=512m; support was removed in 8.0
Scalastyle checks passed.

nssalian commented:
Should have phrased it better, the tests ran fine.

SparkQA commented:
[Test build #1004 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1004/consoleFull) for   PR 7250 at commit [`7621adf`](https://github.com/apache/spark/commit/7621adf368112056eb2e137f62adc429851fa570).

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala:
```diff
@@ -688,6 +690,8 @@ class StreamingContext private[streaming] (
       // The state should always be Stopped after calling `stop()`, even if we haven't started yet
       state = STOPPED
+      // De-registering Streaming Metrics of the StreamingContext
+      env.metricsSystem.removeSource(streamingSource)
```
I think should be done try, after calling scheduler.stop(). In the finally this will fail, if the SparkContext has been stopped and the env is invalid.

jerryshao replied:
I think it would be better to add try as TD said, otherwise you may meet exception when SparkContext is stopped.

nssalian replied:
Changed it on my local. Added to the latest commit.

tdas commented:
You need to add unit tests for this. Please add them to the StreamingContextSuite. You can probably add it to existing unit tests that test start and stop.

SparkQA commented:
[Test build #1004 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1004/console) for   PR 7250 at commit [`7621adf`](https://github.com/apache/spark/commit/7621adf368112056eb2e137f62adc429851fa570).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

nssalian commented:
Will update shortly with the changes.

Thank you.

Commit:
Added test case for de-register metrics and made a change to the scope of the sources ArrayBuffer

nssalian commented:
To illustrate,
Test case:
1) Testing start of the streamingContext and checking state.
2) Storing the size of the sources ArrayBuffer which will have a new source added
3) Sleep for 100 ms.
4) Stopping context and checking state
5) Also checking whether the size of the ArrayBuffer was decreased as the source was removed.
I changed the scope of the sources ArrayBuffer to do this.

Would like some feedback on this approach.

Thank you.

nssalian commented:
@srowen, I wanted to check whether the size was decremented at all.

Couldn't think of a way to assert that the source has been removed since the sources ArrayBuffer is still private.

srowen commented:
But this line doesn't check the state afterwards. You're reading the same value twice; x == x - 1 can never be true. (There's an extra space at the end of that line too.) I think there's still an outstanding comment from TD here too.

Commit:
Added additional variable to check the updated Sources size value to compare with the original size after removal

nssalian commented:
Based on @srowen's comment, I made the change and added updatedSourcesSize to check the ArrayBuffer size after the source is remove to assert that the size was indeed decremented.

jerryshao commented on core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala:
```diff
@@ -73,7 +73,7 @@ private[spark] class MetricsSystem private (
 
   private val sinks = new mutable.ArrayBuffer[Sink]
-  private val sources = new mutable.ArrayBuffer[Source]
+  val sources = new mutable.ArrayBuffer[Source]
```
I think it would be better not to loose the restriction to expose this to the public user just only for test convenience. you could refer [here](https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala#L782) to see how to test with private field.

nssalian replied:
Thanks @jerryshao will add a test similarly.

tdas replied:
@nssalian This has not been addressed. This should not be public and see the link to see how to keep it private and yet access it in tests.

Commit:
Added the removeSource method in try

jerryshao commented on streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala:
```diff
@@ -674,6 +676,8 @@ class StreamingContext private[streaming] (
         case ACTIVE =>
           scheduler.stop(stopGracefully)
+          // De-registering Streaming Metrics of the StreamingContext
+          env.metricsSystem.removeSource(streamingSource)
```
I think here remove the source in `ACTIVE` state may introduce some corner case problem. For example, since metrics source is added when StreamingContext's state is `INITIALIZED`, if we met exception at this point, we will never change the state into `ACTIVE`, so metrics source cannot be removed, since you only assume the state is `ACTIVE` to remove the source. What do you think?

nssalian replied:
The idea was to register at the call of the `start()`.  
So, based on your comment, that would mean registering the sources after the state is set to INITIALIZED and before.

`def start(): Unit = synchronized {
    // Registering Streaming Metrics at the start of the StreamingContext
    assert(env != null)
    assert(env.metricsSystem != null)
    env.metricsSystem.registerSource(streamingSource)`

Makes sense to have it after `INITIALIZED` and before the synchronized block of `ACTIVE` and `STOPPED`.
@tdas, could add more light.

jerryshao replied:
What I mean is to move this line out of `ACTIVE` condition, In any case have to deregister the metrics source, not only in `ACTIVE` state.

tdas replied:
@jerryshao In this PR the streaming metric source is added only after starting the context. So it is fine to deregister only when the state is ACTIVE. Also that maintain the characteristic that `stop` is idempotent.

jerryshao replied:
I think it is OK for normal case, but what if an exception is met after metrics is successfully registered, but before changing the state into `ACTIVE`, according to the code, it will try the exception and change state into `STOPPED`, so at that situation, if we call stop(), we will never de-register the metrics  source according to the current implementation.

tdas replied:
Yeah good point. I suggested above that the registering be moved after the ACTIVE status change.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala:
```diff
@@ -577,6 +575,10 @@ class StreamingContext private[streaming] (
    */
   def start(): Unit = synchronized {
+    // Registering Streaming Metrics at the start of the StreamingContext
+    assert(env != null)
```
This assert can be eliminated.

nssalian replied:
Done.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala:
```diff
@@ -577,6 +575,10 @@ class StreamingContext private[streaming] (
+    // Registering Streaming Metrics at the start of the StreamingContext
+    assert(env != null)
+    assert(env.metricsSystem != null)
+    env.metricsSystem.registerSource(streamingSource)
```
This metrics should be added only inside the case INITIALIZED. The `start` method should be idempotent, and the current form breaks it as calling start will incorrectly attempt to register the source twice.

nssalian replied:
Done. Added The above comment and this into the latest commit.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala:
```diff
@@ -297,6 +296,23 @@ class StreamingContextSuite extends SparkFunSuite with BeforeAndAfter with Timeo
     Thread.sleep(100)
   }
 
+  test("de-register codahale metrics on stop()") {
```
Can you expand the test to cover both registering and deregistering the metrics. so check after start as well as after stop.

Also, rather than count, a more robust check is to find the StreamingSource in MetricSystem.sources. You will have to access private field StreamingContext.streamingSource and MetricSystem.sources.

nssalian replied:
Thanks for the comments. 
Will improve the test and add it in the next commit.

tdas commented:
Also can you update the title to be more specific, similar to the JIRA title.

Commit:
Removed the assert for the env field, added the registerSource line in the INITIALIZED block and kept the removeSource() in the ACTIVE block

Commit:
Added test to check registering and de-registering of streamingSource

Commit:
Removed unused imports

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala:
```diff
@@ -581,6 +579,9 @@ class StreamingContext private[streaming] (
+        // Registering Streaming Metrics at the start of the StreamingContext
+        assert(env.metricsSystem != null)
+        env.metricsSystem.registerSource(streamingSource)
         StreamingContext.ACTIVATION_LOCK.synchronized {
```
This is still a little flawed. there is a case where two StreamingContext can be created, but only one should successfully start. So if one is already started (source registered), the second streaming context.start() should fail. But currently it will fail due to `StreamingContext.assertNoOtherContextIsActive()` but it will attempt to register the source nonetheless. This is bad. So the registering of the source should be done only after that line.

I think it should just before `uiTab.foreach(_.attachI())` line 601

Apologies for changing this so many times.

nssalian replied:
Changed in the latest commit.

Commit:
Moved the registerSource() call before line 601

nssalian commented:
@tdas  made the changes as mentioned.
Does this PR need anything additional/ different?

Thank you.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala:
```diff
@@ -796,3 +815,20 @@ package object testPackage extends Assertions {
+
+/**
+ * Helper methods for testing StreamingContextSuite.
+ * This includes methods to access private methods and fields in ExecutorAllocationManager.
```
Incorrect comment. Please make sure you check everything, comments and all when copying from somewhere else.

nssalian replied:
I figured that would be a good place holder for future methods that may need to be included.
Can re-word accordingly.

tdas replied:
"ExecutorAllocationManager" ??

nssalian replied:
Apologies for missing that. Added in the latest commit with the sources variable and the changing of the comment text.

Commit:
Changed scope of sources and corrected comments for helper

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala:
```diff
@@ -796,3 +815,20 @@ package object testPackage extends Assertions {
+  val streamingSource = getClass.getDeclaredField("streamingSource")
+  streamingSource.setAccessible(true)
+}
+
```
extra empty line

Commit:
Removed extra line in Helper Methods section

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala:
```diff
@@ -297,6 +299,23 @@ class StreamingContextSuite extends SparkFunSuite with BeforeAndAfter with Timeo
+
+    ssc.start()
+    assert(ssc.getState() === StreamingContextState.INITIALIZED)
+    assert(StreamingContextSuite.sources.get(StreamingContextSuite.streamingSource)!= "null")
```
I dont get this statement here. You are calling StreamingContext.sources.get() on an object of type `java.lang.reflect.Field` (that is, StreamingContextSuite.streamingSource) rather than the actual StreamingContext object `ssc` above. And then you are comparing that it is not the string "null"? I am not sure what is being tested here.

nssalian replied:
Checking to see if the source is not returning a null and is actually present in the sources ArrayBuffer.

tdas replied:
A much nicer way to do this is to do it they way it is done in https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala#L782

using the PrivateTester stuff. Please use that, we want to stay consistent to the rest of the code base.

nssalian replied:
Makes sense. I'll write that up. Was having problems when I initially wrote as it was in ExecutionManagerSuite. I'll figure out something.

Thank you.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala:
```diff
@@ -297,6 +299,23 @@ class StreamingContextSuite extends SparkFunSuite with BeforeAndAfter with Timeo
+
+    ssc.stop()
+    assert(ssc.getState() === StreamingContextState.STOPPED)
+    assert(StreamingContextSuite.sources.get(StreamingContextSuite.streamingSource) == "null")
```
Same question here.

Commit:
Added helper method for private methods and changed the test logic to check for Sources containing or not containing StreamingSource

nssalian commented:
@tdas, I added the new test as per PrivateMethodTester.

srowen commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala:
```diff
@@ -21,17 +21,23 @@ import java.io.{File, NotSerializableException}
 import org.apache.spark.util.Utils
-import org.apache.spark.{Logging, SparkConf, SparkContext, SparkException, SparkFunSuite}
+
+import org.apache.spark._
```
I wouldn't make this a _ import. Leave it as it was. ALso I think the collection imports become unnecessary; see below

srowen commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala:
```diff
@@ -297,6 +303,26 @@ class StreamingContextSuite extends SparkFunSuite with BeforeAndAfter with Timeo
+    ssc.start()
+
+    assert(ssc.getState() === StreamingContextState.INITIALIZED)
+    val sources: ArrayBuffer[Source] = StreamingContextSuite.getSources(ssc.env.metricsSystem)
```
I don't think it's necessary to explicitly write the type here. It would be simpler not to.

Commit:
Changed import statements to remove unnecessary imports and add specific imports

Commit:
Removed types for fields in test for registering and deregistering metrics

nssalian commented:
Hitting this error on the test for StreamingContextSuite. 
Is the streamingSource not being found in the right ArrayBuffer? I tried different variations of the registrations to try alleviate this. Didn't help. @tdas  any suggestions?

Ran this:
build/sbt -Pyarn -Phadoop-2.4 -Dhadoop.version=2.5.0 -Pscala-2.10
project streaming
~test-only *StreamingContextSuite

Test failure message:

ArrayBuffer(org.apache.spark.scheduler.DAGSchedulerSource@1473d83a, org.apache.spark.storage.BlockManagerSource@7560392b) did not contain org.apache.spark.streaming.StreamingSource@1c71ddd7 (StreamingContextSuite.scala:322)

srowen commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala:
```diff
@@ -21,17 +21,22 @@ import java.io.{File, NotSerializableException}
 
 import org.apache.commons.io.FileUtils
+import org.apache.spark.metrics.MetricsSystem
+import org.apache.spark.metrics.source.Source
```
These need to be with other spark imports, and the spark imports shouldn't have a blank line in the middle.

srowen commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala:
```diff
@@ -796,3 +821,21 @@ package object testPackage extends Assertions {
+private object StreamingContextSuite extends PrivateMethodTester {
+  private val _sources = PrivateMethod[ArrayBuffer[Source]]('sources)
+  private def getSources(metricsSystem: MetricsSystem): ArrayBuffer[Source] = {
+    metricsSystem invokePrivate _sources()
```
You don't need postfix syntax here; `metricsSystem.invokePrivate(_sources())` is probably better.

Your test fails because the streaming source isn't present after the streaming context stops. Isn't that the point? is your test negated?

Commit:
Changed imports statements, negated test statement and removed postfix

Commit:
Removed added comment to Assert for INITIALIZED state

Commit:
Removing the INITIALIZED check since after start() the state moves to ACTIVE and this check fails

nssalian commented:
The test failed with the  assert(ssc.getState() === StreamingContextState.INITIALIZED)
as after the start() method, the state goes to ACTIVE and fails to match with INITIALIZED.

@srowen, I've added the changes as you suggested.
Thanks.

Commit:
Move the INITIALIZED state check to when the ssc is initialized

srowen commented:
That can't be the failure you're referring to above, right? looks clearly like the assertion about which sources are still in the registered set. Anyway, the assertion looks better now. You need to rebase this PR now, and two of the imports are still in the wrong place.

Commit:
Merge branch 'SPARK-8743' of https://github.com/nssalian/spark into SPARK-8743

nssalian commented:
Right. The negation took care of the failure above and I fixed the assertion error as well.
Removed the spacing for all of them. 

Made a unified commit. 299a57d
Ignore the last one, I''l revert that.

nssalian commented:
Right. The negation took care of the failure above and I fixed the assertion error as well.
Removed the spacing for all of them. 

Made a unified commit. 299a57d
Ignore the last one, I''l revert that.

nssalian commented:
@tdas @srowen, shall I create a fresh PR to avoid any confusion?
I can reference this one.

nssalian commented:
Closing this PR: for #7362

srowen commented:
@nssalian you can make a new PR, yes. I usually like to stick to one PR to preserve the discussion in one place, but it's not a huge deal.

You don't need to squash your commit history like you might in PRs for other projects, since the merge process will take care of that. Still, you can squash some of your commits in your PR branch and force-push if you like for clarity, though you don't need to.

After a rebase to resolve merge conflicts you can simply force-push your PR branch, note.
